src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,780 @@
+
+      if mem == LibC::MAP_FAILED
+        fatal_error "Cannot allocate completion queue: #{Errno.value}"
+      end
+
+      @completion_queue_mmap = mem
+    end
+
+    mem = LibC.mmap(Pointer(Void).null, sq_entries_size,
+      LibC::PROT_READ | LibC::PROT_WRITE, LibC::MAP_SHARED | LibC::MAP_POPULATE,
+      @fd, Syscall::IORING_OFF_SQES)
+
+    if mem == LibC::MAP_FAILED
+      fatal_error "Cannot allocate submission entries: #{Errno.value}"
+    end
+
+    @submission_entries = mem.as(Syscall::IoUringSqe*)
+    @timeval_entries = Slice(LibC::Timeval).new(@params.sq_entries.to_i, make_timeval(::Time::Span::ZERO))
+
+    @submission_queue = SubmissionQueue.new(@fd, @submission_queue_mmap, @params.sq_off)
+    @completion_queue = CompletionQueue.new(@fd, @completion_queue_mmap, @params.cq_off)
+    @closed = false
+  end
+
+  def finalize
+    close unless @closed
+  end
+
+  def close
+    return if @closed
+    @closed = true
+
+    if @submission_queue_mmap && @submission_queue_mmap == @completion_queue_mmap
+      LibC.munmap(@submission_queue_mmap, {sq_size, cq_size}.max)
+    end
+
+    if @submission_queue_mmap
+      LibC.munmap(@submission_queue_mmap, sq_size)
+    end
+
+    if @completion_queue_mmap
+      LibC.munmap(@completion_queue_mmap, cq_size)
+    end
+
+    if @completion_queue_mmap
+      LibC.munmap(@completion_queue_mmap, sq_entries_size)
+    end
+
+    if @fd > 0
+      LibC.close(@fd)
+    end
+  end
+
+  private def sq_size
+    LibC::SizeT.new(@params.sq_off.array + @params.sq_entries * sizeof(UInt32))
+  end
+
+  private def cq_size
+    LibC::SizeT.new(@params.cq_off.cqes + @params.cq_entries * sizeof(Syscall::IoUringCqe))
+  end
+
+  private def sq_entries_size
+    LibC::SizeT.new(@params.sq_entries * sizeof(Syscall::IoUringSqe))
+  end
+
+  private def fatal_error(message)
+    Crystal::System.print_error "\nFATAL (io_uring): #{message}\n"
+    caller.each { |line| Crystal::System.print_error "  from #{line}\n" }
+    exit 1
+  end
+
+  private def make_timeval(time : ::Time::Span)
+    LibC::Timeval.new(
+      tv_sec: LibC::TimeT.new(time.total_seconds),
+      tv_usec: time.nanoseconds // 1_000
+    )
+  end
+
+  # Obtains one free index from the submission queue for our use. If there is none
+  # available then we should call into the kernel to process what is there. Retry
+  # until the kernel has consumed at least one request so that we can write ours.
+  # If there are too many inflight requests, wait for then to complete before sending
+  # more. This check is only needed before the NODROP feature was implemented (Linux 5.5+)
+  private def get_free_index : UInt32
+    # If there are too many in flight events, wait for some completions.
+    until @params.features.nodrop? || @inflight < @params.cq_entries
+      process_completion_events(blocking: true)
+    end
+
+    loop do
+      index = @submission_queue.consume_free_index
+
+      # If the submission queue is full, submit some events.
```

Comment:
So, if I get this straight, this will submit the events when the submission queue is full, and when the scheduler have nothing more to run (using `run_once`).  But what happens if there are fibers running that will always keep runnables around waiting to be scheduled? 

I'm thinking something like 
```
done = false

spawn do
  sleep 1
  done = true
end

spawn do 
  while true
    if done
      exit
    else
      Fiber.yield
    end
  end
end
sleep 100
```

So what I'm wondering is if it needs to submit more eagerly (and lose the lovely batching) or if there needs to be some mechanism in the scheduler to make certain the submit latency don't become too large.

Reply:
That's interesting.

Luckily, `Fiber.yield` is implemented as:

```cr
  protected def yield : Nil
    sleep(0.seconds)
  end
```

And this translates to a NOP io_uring event. So the fiber only executes again after at least one trip to the kernel. It was already like this with libevent (maybe because of the issue you described).

Reply:
Ah, good. I still worry a bit about unforeseen issues that could happen due to potentially unbounded added latency (it may not happen in the Fiber.yield case, but rather in the case the fiber actually execute for a long time). Say a fiber wants to sleep 10 seconds, and then the thread go on and execute for a minute before yielding. Then the thread would be woken up after 1 minute + 10 seconds rather than after 1 minute when the other thread yield. 

But it is perhaps not a case that need to be looked into before anyone actually run into it?